File: libs/power-log-parser/src/lib/cards/triangulate.ts
import { BlockType, GameTag } from '@firestone-hs/reference-data';
import { Action, ShowEntity } from '../models/action';
import { Node, NodeType } from '../models/node';
import type { GameState } from '../state/game-state';
import type { StateFacade } from '../state/state-facade';

/**
 * Discover reveal for Triangulate: nested under PLAY → POWER; deck SpawnToDeck FullEntity closes in a
 * different branch than `node.Parent.Parent === PLAY`, so we walk ancestors for the PLAY block.
 */
function findTriangulateDiscoverShowEntity(
	triggerAction: Action,
	creatorEntityId: number,
): ShowEntity | null {
	const data = triggerAction.GetDataRecursive();
	const linked = data.find(
		(d): d is ShowEntity =>
			d instanceof ShowEntity &&
			!!d.CardId?.length &&
			d.GetTag(GameTag.CREATOR as number) === creatorEntityId,
	);
	if (linked) {
		return linked;
	}
	return (
		data.find(
			(d): d is ShowEntity => d instanceof ShowEntity && !!d.CardId?.length,
		) ?? null
	);
}

export class Triangulate {
	static PredictCardId(
		gameState: GameState,
		creatorCardId: string,
		creatorEntityId: number,
		node: Node,
		stateFacade: StateFacade | null,
	): string | null {
		let playAction: Action | null = null;
		let cur: Node | null = node;
		while (cur != null) {
			if (cur.Type === NodeType.Action) {
				const act = cur.Object as Action;
				if (act.Type === (BlockType.PLAY as number) && act.Entity === creatorEntityId) {
					playAction = act;
					break;
				}
			}
			cur = cur.Parent;
		}

		if (!playAction) {
			return null;
		}

		const triggerAction = playAction.Data.filter((d): d is Action => d instanceof Action).find(
			(a) => a.Type === (BlockType.POWER as number) && a.Entity === creatorEntityId,
		);
		if (!triggerAction) {
			return null;
		}

		const showEntity = findTriangulateDiscoverShowEntity(triggerAction, creatorEntityId);
		if (!showEntity) {
			return null;
		}

		const canReveal =
			showEntity.GetTag(GameTag.CASTS_WHEN_DRAWN) === 1 ||
			showEntity.GetTag(GameTag.SUMMONED_WHEN_DRAWN) === 1;
		const lp = stateFacade?.LocalPlayer;
		const triCaster =
			stateFacade?.GsState?.GameState.CurrentEntities.get(creatorEntityId) ??
			gameState.CurrentEntities.get(creatorEntityId);
		const casterController = triCaster?.GetEffectiveController() ?? -1;
		// Spectator logs: hide opponent discover. Ladder / own power.log: LocalPlayer heuristics can mismatch
		// the real account; do not gate on LocalPlayer there (see triangulate-created-by replay spec).
		if (
			stateFacade?.Spectating &&
			lp != null &&
			casterController !== lp.PlayerId &&
			!canReveal
		) {
			return null;
		}

		return showEntity.CardId ?? null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Cards/Triangulate.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Entities;

namespace HearthstoneReplays.Events.Cards
{
    internal class Triangulate
    {
        public static string PredctCardId(GameState gameState, string creatorCardId, int creatorEntityId, Node node, StateFacade stateFacade)
        {
            if (node.Parent?.Type != typeof(Action))
            {
                return null;
            }
            
            var act = node.Parent.Parent.Object as Action;
            if (act.Type != (int)BlockType.PLAY)
            {
                return null;
            }

            var actionEntity = gameState.CurrentEntities.GetValueOrDefault(act.Entity);
            var controller = actionEntity.GetController();

            var triggerAction = act.Data
                .Where(d => d is Action)
                .Select(d => d as Action)
                .Where(a => a.Type == (int)BlockType.POWER && a.Entity == act.Entity)
                .FirstOrDefault();
            if (triggerAction == null)
            {
                return null;
            }

            // First ShowEntity is the card we've picked?
            var showEntity = triggerAction.Data
                .Where(d => d is ShowEntity)
                .Select(d => d as ShowEntity)
                .FirstOrDefault();

            // Avoid info leaks, as the card is revealed in the logs
            // UNLESS the card is CAST_WHEN_DRAWN / SUMMONED_WHEN_DRAWN
            var canReveal = showEntity.GetTag(GameTag.CASTS_WHEN_DRAWN) == 1 || showEntity.GetTag(GameTag.SUMMONED_WHEN_DRAWN) == 1;
            if (controller != stateFacade.LocalPlayer.PlayerId && !canReveal)
            {
                return null;
            }

            return showEntity?.CardId;
        }
    }
}
